### PR TITLE
[WIP] Fixes in Simple DQN.

### PR DESCRIPTION
**Pull request by kartikdutt18**
These are some simple changes in simple DQN. The aim of this PR is to resolve invalid memory fixes.

Current changes:
1. Use pass by reference.
2. Some minor style change.

There are other things that need to be fixed
1. Use templated networks (that can be changed, I kind of explained this in my review below).
2. Run valgrind a couple of times till this is fixed.
3. Is lack of Copy Constructors in other layers causing an issue?

Let me know if I missed something.

### CODE REVIEW

**Comment by kartikdutt18** [comment]
So the issue I was talking about above (hope this isn't too long).
1. The line [here](https://github.com/mlpack/mlpack/blob/1c25a1bda52832841efec4e41477bfcfbbbf6f3f/src/mlpack/methods/reinforcement_learning/q_networks/simple_dqn.hpp#L53) doesn't match the template type. 
2. Also we are making an unnecessary copy [here](https://github.com/mlpack/mlpack/blob/1c25a1bda52832841efec4e41477bfcfbbbf6f3f/src/mlpack/methods/reinforcement_learning/q_networks/simple_dqn.hpp#L60). Not sure but if we avoid copying maybe that could fix the issue.

A solution to this is creating a the class to template parameters i.e. OutputLayerType and InitializationRuleType. Similar to what I have done in mlpack/models#3. Let me know what you think about this and I'll make the changes in the subsequent commits.

**Comment by kartikdutt18** [comment]
Hey @zoq, Are there any visualizers for the environments?
I can't seem to find any in reinforcement learning folder.

**Comment by zoq** [comment]
You could use https://github.com/zoq/gym_tcp_api.

File: src/mlpack/methods/reinforcement_learning/q_networks/simple_dqn.hpp [hunk 1]
```diff
@@ -110,8 +113,8 @@ class SimpleDQN
    * @param target The training target.
    * @return gradient The gradient.
    */
-  void Backward(const arma::mat state, arma::mat& target,
-arma::mat& gradient)
+  void Backward(const arma::mat& state, arma::mat& target,
+      arma::mat& gradient)
```
Comment (zoq) [inline src/mlpack/methods/reinforcement_learning/q_networks/simple_dqn.hpp:117]: Can you put each parameter on a seperate line?

**Comment by zoq** [comment]
This is still work in progress right?

**Comment by kartikdutt18** [comment]
> This is still work in progress right?

Yes. Sorry you had to review this when it wasn't fully ready. I'll add WIP to the name of the PR. Also sorry that I have been a bit slow with my PR lately, once my online assessment and college end on 6th May, I should be able to wrap them up much faster. I'll try completing as much as I can during the spare time I get till then. Thanks a lot.

**Comment by zoq** [comment]
No worries at all, just wanted to make sure this is WIP, hopefully the online assessment is a little bit of fun.

**Comment by kartikdutt18** [comment]
> hopefully the online assessment is a little bit of fun.

Haha, It is definitely a lot better than offline exams.

**Pull request closed without merging**